hudi-spark/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala

```diff
@@ -0,0 +1,209 @@
+      sparkSession = sqlContext.sparkSession,
+      dataSchema = tableStructSchema,
+      partitionSchema = StructType(Nil),
+      requiredSchema = tableStructSchema,
+      filters = pushDownFilter,
+      options = optParams,
+      hadoopConf = sqlContext.sparkSession.sessionState.newHadoopConf()
+    )
+    val requiredSchemaParquetReader = new ParquetFileFormat().buildReaderWithPartitionValues(
+      sparkSession = sqlContext.sparkSession,
+      dataSchema = tableStructSchema,
+      partitionSchema = StructType(Nil),
+      requiredSchema = requiredStructSchema,
+      filters = pushDownFilter,
+      options = optParams,
+      hadoopConf = sqlContext.sparkSession.sessionState.newHadoopConf()
+    )
+
+    // Follow the implementation of Spark internal HadoopRDD to handle the broadcast configuration.
+    FileSystem.getLocal(jobConf)
+    SparkHadoopUtil.get.addCredentials(jobConf)
+    val rdd = new HoodieMergeOnReadRDD(
+      sqlContext.sparkContext,
+      jobConf,
+      fullSchemaParquetReader,
+      requiredSchemaParquetReader,
+      hoodieTableState
+    )
+    rdd.asInstanceOf[RDD[Row]]
+  }
+
+  def buildFileIndex(): List[HoodieMergeOnReadFileSplit] = {
+    val affectedFileStatus = new ListBuffer[FileStatus]
+    for (commit <- commitsToReturn) {
+      val metadata: HoodieCommitMetadata = HoodieCommitMetadata.fromBytes(commitsTimelineToReturn.getInstantDetails(commit)
+        .get, classOf[HoodieCommitMetadata])
+      val idWithPath = metadata.getFileIdAndFullPaths(metaClient.getBasePath).toMap
+      idWithPath.foreach(p => {
+        val file = fs.getFileStatus(new Path(p._2))
+        affectedFileStatus += file
+      })
+    }
+    val fsView = new HoodieTableFileSystemView(metaClient,
+      commitsTimelineToReturn, affectedFileStatus.toArray)
+    val fileGroup = fsView.fetchAllStoredFileGroups().iterator().toList
+    val latestCommit = fsView.getLastInstant.get().getTimestamp
+    if (log.isDebugEnabled) {
+      fileGroup.foreach(f => log.debug(s"current file group id: " +
+        s"${f.getFileGroupId} and file slices ${f.getLatestFileSlice.get().toString}"))
+    }
+    val pathGlobPattern = optParams.getOrElse(
+      DataSourceReadOptions.INCR_PATH_GLOB_OPT_KEY,
+      DataSourceReadOptions.DEFAULT_INCR_PATH_GLOB_OPT_VAL)
+    val filteredFileGroup = if(!pathGlobPattern
+      .equals(DataSourceReadOptions.DEFAULT_INCR_PATH_GLOB_OPT_VAL)) {
+      val globMatcher = new GlobPattern("*" + pathGlobPattern)
+      fileGroup.filter(f => {
+        if (f.getLatestFileSlice.get().getBaseFile.isPresent) {
+          globMatcher.matches(f.getLatestFileSlice.get().getBaseFile.get.getPath)
+        } else {
+          globMatcher.matches(f.getLatestFileSlice.get().getLatestLogFile.get().getPath.toString)
+        }
+      })
+    } else {
+      fileGroup
+    }
+
+    filteredFileGroup.map(f => {
+      val baseFile = f.getLatestDataFile
```

Comment:
IIUC, your approach here is to find all the file groups impacted in commit range, then pull affected records from the latest file slice? the mergeOnReadRDD will handle the merging of such an file slice, with the commit filters applied. 

let me know @garyli1019 if I am understanding this correctly. it will help me review the code. 

in the meantime, can we add a separate flag/option to turn on `new.incremental.relation=true` or sth, to control this and push the changes

Reply:
correct, that's what I was trying to do here.
sure, keep only one incremental query type makes sense to me.

Reply:
@garyli1019 one issue we need to make sure we handle is when the file group is pending compaction. when that happens, the base file is not present in the latest slice, but stitched together from the previous slice. @bhasudha is handling this in the Hive PR. worth taking a look and ensuring

Reply:
I think we can share the implementation of this part for Hive and Spark. I will move this part to a util class.

Reply:
Just to make sure. This only possible for the base file right? The latest log file will always be on the latest file slice?

Reply:
@vinothchandar getting back to this PR... Now I get a little confused, so when we pull from a pending compaction commit, why should we care about the base file that was not in the range? We just read the log file right? From a user perspective, I don't expect to read commit earlier than the commit I defined as a starting point.